File: sdk/AuthSys-Device-Example/ruby/AuthSys-Device.rb
require 'net/http'
require 'json'
require 'digest'
require 'socket'

class AuthSysDevice
  attr_reader :last_error, :last_response

  def initialize(group_secret, base_url = 'https://api.authsys.dpdns.org/api/v1/client')
    @group_secret = group_secret
    @base_url = base_url.sub(/\/+$/, '')
    @last_error = ''
    @last_response = ''
    @uri = URI(@base_url)
  end

  def self.get_hwid
    raw = Socket.gethostname
    if RUBY_PLATFORM =~ /mingw|mswin/
      begin
        raw += `wmic nic where "NetEnabled=true" get MACAddress 2>nul`.to_s
      rescue; end
    else
      begin
        raw += `ifconfig 2>/dev/null`.to_s
      rescue; end
    end
    Digest::MD5.hexdigest(raw).upcase
  rescue
    'unknown'
  end

  private

  def request(endpoint, payload)
    uri = URI("#{@base_url}/#{endpoint}")
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == 'https')
    http.open_timeout = 15
    http.read_timeout = 15

    req = Net::HTTP::Post.new(uri.path)
    req['Content-Type'] = 'application/json'
    req.body = payload.to_json

    res = http.request(req)
    @last_response = res.body
    JSON.parse(res.body)
  rescue => e
    @last_error = e.message
    nil
  end

  public

  def check
    @last_error = ''
    data = request('check', {
      group_secret: @group_secret,
      hwid: self.class.get_hwid
    })
    return false unless data

    if data['active'] == true
      true
    else
      @last_error = data['message'] || 'Device deactivated by admin'
      false
    end
  end

  def register(device_name = '')
    @last_error = ''
    payload = { group_secret: @group_secret, hwid: self.class.get_hwid }
    payload[:device_name] = device_name unless device_name.empty?

    data = request('register', payload)
    return false unless data

    data['active'] == true
  end
end

File: sdk/AuthSys-Device-Example/php/AuthSys-Device.php
<?php

class AuthSysDevice
{
    private string $appSecret;
    private string $baseUrl;
    public string $lastError = '';
    public string $lastResponse = '';

    public function __construct(string $appSecret, string $baseUrl = 'https://authsys-main-production.up.railway.app/device')
    {
        $this->appSecret = $appSecret;
        $this->baseUrl = rtrim($baseUrl, '/');
    }

    private static function getHWID(): string
    {
        if (PHP_OS_FAMILY === 'Windows') {
            $output = shell_exec('wmic bios get serialnumber 2>nul');
            if ($output) {
                $lines = explode("\n", trim($output));
                return isset($lines[1]) ? trim($lines[1]) : 'unknown';
            }
        } else {
            $id = @file_get_contents('/etc/machine-id');
            if ($id) return trim($id);
        }
        return 'unknown';
    }

    private function request(string $endpoint, array $payload): array
    {
        $url = $this->baseUrl . '/' . $endpoint;
        $json = json_encode($payload);

        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $json,
            CURLOPT_HTTPHEADER => ['Content-Type: application/json'],
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 15,
            CURLOPT_SSL_VERIFYPEER => false,
        ]);

        $this->lastResponse = curl_exec($ch) ?: '';
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        $data = json_decode($this->lastResponse, true);
        return $data ?: [];
    }

    public function check(): bool
    {
        $this->lastError = '';
        $data = $this->request('check', [
            'device_key' => $this->appSecret,
            'hwid' => self::getHWID(),
        ]);

        if (!empty($data['active'])) return true;
        $this->lastError = $data['message'] ?? 'Device deactivated by admin';
        return false;
    }

    public function register(string $deviceName = ''): bool
    {
        $this->lastError = '';
        $payload = [
            'device_key' => $this->appSecret,
            'hwid' => self::getHWID(),
        ];
        if ($deviceName) $payload['device_name'] = $deviceName;

        $data = $this->request('register', $payload);
        return !empty($data['active']);
    }
}
